src/tracker/tracker_udp.cc

```diff
@@ -70,25 +70,22 @@ TrackerUdp::TrackerUdp(TrackerList* parent, const std::string& url, int flags) :
 
   m_port(0),
 
-  m_slot_resolver(NULL),
   m_readBuffer(NULL),
   m_writeBuffer(NULL) {
 
   m_taskTimeout.slot() = std::bind(&TrackerUdp::receive_timeout, this);
+
+  m_resolver_callback = std::bind(&TrackerUdp::start_announce, this, std::placeholders::_1, std::placeholders::_2);
```

Comment:
This is not stable. We can't have the callback in the `TrackerUdp` object. If the torrent is removed before the announce completes, it will trigger a crash. The callback will no longer be valid.

Reply:
Just to be clear, this line hasn't changed in years. You've been deploying this and it's been running stably, correct?

Reply:
Also if I understand your concern correctly, it's addressed by having `close_directly` remove the callback:

https://github.com/rakshasa/libtorrent/blob/a03859ee9fa9f65beeaec01c7694f8c8959568fa/src/tracker/tracker_udp.cc#L180-L182

Reply:
Nope, we passing are `m_resolver_callback` as a pointer to enqueue. This must remain valid indefinitely because it's async. It was stable on my test environment. But once we tested it with 10k of clients, it started flooding crash dumps.
```c
  m_resolver_query = manager->connection_manager()->async_resolver().enqueue(
      hostname.data(),
      AF_INET,
      &m_resolver_callback
  );
```

This is the solution I implemented to avoid sending a pointer that could potentially become undefined. 
https://github.com/stickz/rtorrent/compare/32e1e6f6357d35b53aea947da16d9526fc5a3895...5517bef3cc71f1ac6191b89ae1194080385ddb18

Reply:
The callbacks are "asynchronous" in the sense that the main event loop does not block while the query is in flight, but the callbacks can only be triggered from `::dns_ioevent` or `::dns_timeouts`, which are called from within the main event loop (once the poll layer detects that a response is ready, or that the query has failed). Accordingly a call to `::dns_cancel` from the main event loop can prevent the callbacks from triggering. Because the main loop runs within a single thread, it should not be possible for these operations to race against each other.

According to my understanding, the specific scenario you described (DNS query is initiated, the torrent is removed, DNS response is received, the callback executes against a freed `TrackerUdp` object) cannot occur for this exact reason: `TrackerUdp::~TrackerUdp` calls `UdnsAsyncResolver::cancel`, which calls `::dns_cancel` on the callbacks pointing to the `TrackerUdp` object that is being freed.

I respect the intensive testing you have done on this patchset but I have run it continuously for years on heavy workloads without issue. I have also tested it with Valgrind, which found memory safety issues in the DHT layer (e.g. #166) but not with the udns code. I think it's possible that the issues you observed are a bad interaction with another patchset you're using? I would also be interested to see any backtraces or stack dumps you were able to collect.

Reply:
Here's a stack trace I was able to collect and resolve for this problem.

```
Thread 1 (Thread 0x7f04816ad680 (LWP 114901)):
#0  0x00007f04820e350b in pthread_kill () from /usr/lib64/libc.so.6
#1  0x00007f0482086906 in raise () from /usr/lib64/libc.so.6
#2  0x00007f04820698b7 in abort () from /usr/lib64/libc.so.6
#3  0x000055de99b537c8 in do_panic (signum=11) at /usr/src/debug/net-p2p/rtorrent-0.9.8.5.4/rtorrent-5.4-0.9.8-0.13.8/rtorrent/src/main.cc:645
#4  <signal handler called>
#5  0x000055de9dd38e40 in ?? ()
#6  0x00007f0482613280 in std::function<void (sockaddr const*, int)>::operator()(sockaddr const*, int) const (__args#1=<optimized out>, __args#0=<optimized out>, this=<optimized out>, this=<optimized out>, __args#0=<optimized out>, __args#1=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/std_function.h:591
#7  torrent::a4_callback_wrapper (result=<optimized out>, data=0x7f04541fc480, ctx=<optimized out>) at utils/udnsevent.cc:60
#8  torrent::a4_callback_wrapper (ctx=<optimized out>, result=<optimized out>, data=0x7f04541fc480) at utils/udnsevent.cc:40
#9  0x00007f0481addf1c in dns_ioevent () from /usr/lib64/libudns.so.0
#10 0x00007f04825a9e9e in torrent::PollEPoll::perform (this=0x55de9ac6b630) at torrent/poll_epoll.cc:185
#11 0x00007f04825d41b4 in torrent::thread_base::event_loop (thread=0x55de9ac72300) at torrent/utils/thread_base.cc:174
#12 0x000055de99b7aa5c in main (argc=<optimized out>, argv=<optimized out>) at /usr/src/debug/net-p2p/rtorrent-0.9.8.5.4/rtorrent-5.4-0.9.8-0.13.8/rtorrent/src/main.cc:508
```

Reply:
Thanks for getting back to me on this.

I believe this is a bug in your fork. You originally merged the udns changes in https://github.com/stickz/rtorrent/commit/62d3f5e2a81a666f8b708a740142339b0557ff04 ; this version of the patch correctly includes the call to `AsyncResolver::cancel` in `TrackerUdp::close_directly` that I linked above. Then subsequently in https://github.com/stickz/rtorrent/commit/a843234b54372fd89166ad75a737a4e5f79f4284 you replaced `tracker_udp.cc` with a version (from libtorrent's master?) that does not include this call. Consequently, the callback does not get cancelled and can fire even when it has become invalid, which is what we see in the stack trace.